Source/WebCore/style/StyleInvalidator.cpp

```diff
@@ -345,29 +345,25 @@ void Invalidator::invalidateStyleWithMatchElement(Element& element, MatchElement
         break;
     }
     case MatchElement::HasSibling: {
-        if (auto* sibling = element.previousElementSibling()) {
-            SelectorMatchingState selectorMatchingState;
-            if (RefPtr parent = element.parentElement())
-                selectorMatchingState.selectorFilter.pushParentInitializingIfNeeded(*parent);
-
-            for (; sibling; sibling = sibling->previousElementSibling())
-                invalidateIfNeeded(*sibling, &selectorMatchingState);
-        }
+        SelectorMatchingState selectorMatchingState;
+        if (auto* parent = element.parentElement())
+            selectorMatchingState.selectorFilter.pushParentInitializingIfNeeded(*parent);
+        for (auto& sibling : childrenOfType<Element>(*element.parentNode()))
+            invalidateIfNeeded(sibling, &selectorMatchingState);
```

Comment:
Maybe we should add another MatchElement for this scope breaking sibling case? It is not great that the common case :has(~ foo) gets less efficient to cover this obscure corner.